README.md
# ./fabio [![Build Status](https://travis-ci.org/eBay/fabio.svg?branch=master)](https://travis-ci.org/eBay/fabio)

##### Current version: 1.0.3

fabio is a fast, modern, zero-conf load balancing HTTP(S) router for deploying
microservices managed by consul.

It provides a single-binary alternative to running [consul-
template](https://github.com/hashicorp/consul-template) together with
haproxy/varnish/nginx/apache. Services provide one or more host/path prefixes
they serve and fabio updates the routing table every time a service becomes
(un-)available without restart.

fabio was developed at the [eBay Classifieds Group](http://www.ebayclassifiedsgroup.com)
in Amsterdam and is currently used to route traffic for
[marktplaats.nl](http://www.marktplaats.nl) and [kijiji.it](http://www.kijiji.it).
Marktplaats is running all of its traffic through fabio which is
several thousand requests per second distributed over several fabio
instances.

## Features

* Single binary in Go. No external dependencies.
* Zero-conf
* Hot-reloading of routing table through backend watchers
* Round robin and random distribution
* [Traffic Shaping](#Traffic Shaping) (send 5% of traffic to new instances)
* SSL client certificate authentication support
* Graphite metrics
* Request tracing
* WebUI
* Fast

## Documentation

* [Installation](#installation)
* [Quickstart](#quickstart)
* [Configuration](https://raw.githubusercontent.com/eBay/fabio/master/fabio.properties) (documented fabio.properties file)
* [Performance](#performance)
* [Service configuration](#service-configuration)
* [Manual overrides](#manual-overrides)
* [Routing](#routing)
* [Traffic shaping](#traffic-shaping)
* [Debugging](#debugging)
* [Request tracing](#request-tracing)
* [Web UI](#web-ui)
* [Changelog](https://github.com/eBay/fabio/blob/master/CHANGELOG.md)
* [License](#license)

## Quickstart

This is how you use fabio in your setup:

1. Register your service in consul
2. Register a **health check** in consul as described [here](https://consul.io/docs/agent/checks.html).
   Make sure the health check is **passing** since fabio will only watch services
   which have a passing health check.
3. Register one `urlprefix-` tag per `host/path` prefix it serves,
   e.g. `urlprefix-/css`, `urlprefix-/static`, `urlprefix-mysite.com/`
4. Start fabio without a config file (assuming a consul agent on `localhost:8500`)
   Watch the log output how fabio picks up the route to your service.
   Try starting/stopping your service to see how the routing table changes instantly.
5. Send all your HTTP traffic to fabio on port `9999`
6. Done

If you want fabio to handle SSL as well set the `proxy.addr` along with the
public/private key files in
[fabio.properties](https://github.com/eBay/fabio/blob/master/fabio.properties)
and run `fabio -cfg fabio.properties`. You might also want to set the
`proxy.header.clientip`, `proxy.header.tls` and `proxy.header.tls.value`
options.

Check the [Debugging](#debugging) section to see how to test fabio with `curl`.

See fabio in action

[![fabio demo](http://i.imgur.com/aivFAKl.png)](https://www.youtube.com/watch?v=gvxxu0PLevs"fabio demo - Click to Watch!")

## Installation

To install fabio run (you need Go 1.4 or higher)

    go get github.com/eBay/fabio

To start fabio run

    ./fabio

which will run it with the default configuration which is described
in `fabio.properties`. To run it with a config file run it
with

    ./fabio -cfg fabio.properties

or use the official Docker image and mount your own config file to `/etc/fabio/fabio.properties`

    docker run -d -p 9999:9999 -p 9998:9998 -v $PWD/fabio/fabio.properties:/etc/fabio/fabio.properties magiconair/fabio

If you want to run the Docker image with one or more SSL certificates then
you can store your configuration and certificates in `/etc/fabio` and mount
the entire directory, e.g.

    $ cat ~/fabio/fabio.properties
    proxy.addr=:443;/etc/fabio/ssl/mycert.pem;/etc/fabio/ssl/mykey.pem

    docker run -d -p 443:443 -p 9998:9998 -v $PWD/fabio:/etc/fabio magiconair/fabio

The official Docker image contains the root CA certificates from a recent and updated
Ubuntu 12.04.5 LTS installation.

## Performance

fabio has been tested to deliver up to 15.000 req/sec on a single 16
core host with moderate memory requirements (~ 60 MB).

To achieve the performance fabio sets the following defaults which
can be overwritten with the environment variables:

* `GOMAXPROCS` is set to `runtime.NumCPU()` since this is not the
  default for Go 1.4 and before
* `GOGC=800` is set to reduce the pressure on the garbage collector

When fabio is compiled with Go 1.5 and run with default settings it can be up
to 40% slower  than the same version compiled with Go 1.4. The `GOGC=100`
default puts more pressure on the Go 1.5 GC which makes the fabio spend 10% of
the time in the GC. With `GOGC=800` this drops back to 1-2%. Higher values
don't provide higher gains.

As usual, don't rely on these numbers and perform your own benchmarks. You can
check the time fabio spends in the GC with `GODEBUG=gctrace=1`.

## Service configuration

Each service can register one or more URL prefixes for which it serves
traffic. A URL prefix is a `host/path` combination without a scheme since SSL
has already been terminated and all traffic is expected to be HTTP. To
register a URL prefix add a tag `urlprefix-host/path` to the service
definition.

By default, traffic is distributed evenly across all service instances which
register a URL prefix but you can set the amount of traffic a set of instances
will receive ("Canary testing"). See [Traffic Shaping](#Traffic Shaping)
below.

A background process watches for service definition and health status changes
in consul. When a change is detected a new routing table is constructed using
the commands described in [Config Commands](#Config Commands).

## Manual overrides

Since an automatically generated routing table can only be changed with a
service deployment additional routing commands can be stored manually in the
consul KV store which get appended to the automatically generated routing
table. This allows fine-tuning and fixing of problems without a deployment.

The [Traffic Shaping](#Traffic Shaping) commands are also stored in the KV
store.

## Routing Table Configuration

The routing table is configured with the following commands:

```
route add service host/path targetURL [weight <weight>] [tags "tag1,tag2,..."]
	- Add a new route for host/path to targetURL

route del service
	- Remove all routes for service

route del service host/path
	- Remove all routes for host/path for this service only

route del service host/path targetURL
	- Remove only this route

route weight service host/path weight n tags "tag1,tag2"
  - Route n% of traffic to services matching service, host/path and tags
    n is a float > 0 describing a percentage, e.g. 0.5 == 50%
    n <= 0: means no fixed weighting. Traffic is evenly distributed
    n > 0: route will receive n% of traffic. If sum(n) > 1 then n is normalized.
    sum(n) >= 1: only matching services will receive traffic

```

The order of commands matters but routes are always ordered from most to least
specific by prefix length.

## Routing

The routing table contains first all routes with a host sorted by prefix
length in descending order and then all routes without a host again sorted by
prefix length in descending order.

For each incoming request the routing table is searched top to bottom for a
matching route. A route matches if either `host/path` or - if there was no
match - just `/path` matches.

The matching route determines the target URL depending on the configured
strategy. `rnd` and `rr` are available with `rnd` being the default.

### Example

The auto-generated routing table is

```
route add service-a www.mp.dev/accounts/ http://host-a:11050/ tags "a,b"
route add service-a www.kjca.dev/accounts/ http://host-a:11050/ tags "a,b"
route add service-a www.dba.dev/accounts/ http://host-a:11050/ tags "a,b"
route add service-b www.mp.dev/auth/ http://host-b:11080/ tags "a,b"
route add service-b www.kjca.dev/auth/ http://host-b:11080/ tags "a,b"
route add service-b www.dba.dev/auth/ http://host-b:11080/ tags "a,b"
```

The manual configuration under `/fabio/config` is

```
route del service-b www.dba.dev/auth/
route add service-c www.somedomain.com/ http://host-z:12345/
```

The complete routing table then is

```
route add service-a www.mp.dev/accounts/ http://host-a:11050/ tags "a,b"
route add service-a www.kjca.dev/accounts/ http://host-a:11050/ tags "a,b"
route add service-a www.dba.dev/accounts/ http://host-a:11050/ tags "a,b"
route add service-b www.mp.dev/auth/ http://host-b:11080/ tags "a,b"
route add service-b www.kjca.dev/auth/ http://host-b:11080/ tags "a,b"
route add service-c www.somedomain.com/ http://host-z:12345/ tags "a,b"
```

## Traffic Shaping

fabio allows to control the amount of traffic a set of service
instances will receive. You can use this feature to direct a fixed percentage
of traffic to a newer version of an existing service for testing ("Canary
testing").

The following command will allocate 5% of traffic to `www.kjca.dev/auth/` to
all instances of `service-b` which match tags `version-15` and `dc-fra`. This
is independent of the number of actual instances running. The remaining 95%
of the traffic will be distributed evenly across the remaining instances
publishing the same prefix.

```
route weight service-b www.kjca.dev/auth/ weight 0.05 tags "version-15,dc-fra"
```

## Debugging

To send a request from the command line via the fabio using `curl`
you should send it as follows:

```
curl -v -H 'Host: foo.com' 'http://localhost:9999/path'
```

The `-x` or `--proxy` options will most likely not work as you expect as they
send the full URL instead of just the request URI which usually does not match
any route but the default one - if configured.

## Request tracing

To trace how a request is routed you can add a `Trace` header with an non-
empty value which is truncated at 16 characters to keep the log output short.

```
$ curl -v -H 'Trace: abc' -H 'Host: foo.com' 'http://localhost:9999/bar/baz'

2015/09/28 21:56:26 [TRACE] abc Tracing foo.com/bar/baz
2015/09/28 21:56:26 [TRACE] abc No match foo.com/bang
2015/09/28 21:56:26 [TRACE] abc Match foo.com/
2015/09/28 22:01:34 [TRACE] abc Routing to http://1.2.3.4:8080/
```

## Web UI

fabio contains a (very) simple web ui to examine the routing
table. By default it is accessible on `http://localhost:9998/`

## License

MIT licensed



CHANGELOG.md
## Changelog

### ???

 * Add support for SSL client certificate authentication

### [v1.0.3](https://github.com/eBay/fabio/releases/tag/v1.0.3) - 25 Oct 2015

 * Add Docker support and official Docker image `magiconair/fabio`

### [v1.0.2](https://github.com/eBay/fabio/releases/tag/v1.0.2) - 23 Oct 2015

 * [Pull Request #3](https://github.com/eBay/fabio/pull/3): Honor consul.url and consul.addr from config file (@jeinwag)

### [v1.0.1](https://github.com/eBay/fabio/releases/tag/v1.0.1) - 21 Oct 2015

 * Honor maintenance mode for both services and nodes

### [v1.0.0](https://github.com/eBay/fabio/releases/tag/v1.0.0) - 16 Oct 2015

 * Initial open-source release



listen_test.go
package main

import (
	"net/http"
	"net/http/httptest"
	"sync"
	"testing"
	"time"

	"github.com/eBay/fabio/route"
)

// TestGracefulShutdown tests
func TestGracefulShutdown(t *testing.T) {

	req := func(url string) int {
		resp, err := http.Get(url)
		if err != nil {
			t.Fatal(err)
		}
		defer resp.Body.Close()
		return resp.StatusCode
	}

	// start a server which responds after the shutdown has been triggered.
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		<-quit // wait for shutdown signal
		return
	}))
	defer srv.Close()

	// load the routing table
	tbl, err := route.ParseString("route add svc / " + srv.URL)
	if err != nil {
		t.Fatal(err)
	}
	route.SetTable(tbl)

	// start proxy with graceful shutdown period long enough
	// to complete one more request.
	var wg sync.WaitGroup
	wg.Add(1)
	laddr := "127.0.0.1:57777"
	go func() {
		defer wg.Done()
		listen(laddr, 250*time.Millisecond, route.NewProxy(http.DefaultTransport, "", "", "", ""))
	}()

	// trigger shutdown after some time
	shutdownDelay := 100 * time.Millisecond
	go func() {
		time.Sleep(shutdownDelay)
		close(quit)
	}()

	// give proxy some time to start up
	// needs to be done before shutdown is triggered
	time.Sleep(shutdownDelay / 2)

	// make 200 OK request
	// start before and complete after shutdown was triggered
	if got, want := req("http://"+laddr+"/"), 200; got != want {
		t.Fatalf("request 1: got %v want %v", got, want)
	}

	// make 503 request
	// start and complete after shutdown was triggered
	if got, want := req("http://"+laddr+"/"), 503; got != want {
		t.Fatalf("got %v want %v", got, want)
	}

	// wait for listen() to return
	// note that the actual listeners have not returned yet
	wg.Wait()
}


route/proxy_test.go
package route

import (
	"crypto/tls"
	"net/http"
	"reflect"
	"testing"
)

func TestAddHeaders(t *testing.T) {
	tests := []struct {
		r              *http.Request
		localIP        string
		remoteIPHeader string
		tlsHeader      string
		tlsHeaderValue string
		hdrs           http.Header
		err            string
	}{
		{ // error
			&http.Request{RemoteAddr: "1.2.3.4"},
			"", "", "", "",
			http.Header{},
			"cannot parse 1.2.3.4",
		},

		{ // set remote ip header
			&http.Request{RemoteAddr: "1.2.3.4:5555"},
			"", "Client-IP", "", "",
			http.Header{"Forwarded": {"for=1.2.3.4; proto=http"}, "Client-Ip": []string{"1.2.3.4"}},
			"",
		},

		{ // set remote ip header with local ip (no change expected)
			&http.Request{RemoteAddr: "1.2.3.4:5555"},
			"5.6.7.8", "Client-IP", "", "",
			http.Header{"Forwarded": {"for=1.2.3.4; proto=http; by=5.6.7.8"}, "Client-Ip": []string{"1.2.3.4"}},
			"",
		},

		{ // set X-Forwarded-For
			&http.Request{RemoteAddr: "1.2.3.4:5555"},
			"", "X-Forwarded-For", "", "",
			http.Header{"Forwarded": {"for=1.2.3.4; proto=http"}, "X-Forwarded-For": []string{"1.2.3.4"}},
			"",
		},

		{ // set X-Forwarded-For with local ip
			&http.Request{RemoteAddr: "1.2.3.4:5555"},
			"5.6.7.8", "X-Forwarded-For", "", "",
			http.Header{"Forwarded": {"for=1.2.3.4; proto=http; by=5.6.7.8"}, "X-Forwarded-For": []string{"1.2.3.4, 5.6.7.8"}},
			"",
		},

		{ // extend X-Forwarded-For with local ip
			&http.Request{RemoteAddr: "1.2.3.4:5555", Header: http.Header{"X-Forwarded-For": []string{"9.9.9.9"}}},
			"5.6.7.8", "", "", "",
			http.Header{"Forwarded": {"for=1.2.3.4; proto=http; by=5.6.7.8"}, "X-Forwarded-For": []string{"9.9.9.9, 5.6.7.8"}},
			"",
		},

		{ // set Forwarded
			&http.Request{RemoteAddr: "1.2.3.4:5555"},
			"", "", "", "",
			http.Header{"Forwarded": {"for=1.2.3.4; proto=http"}},
			"",
		},

		{ // set Forwarded with localIP
			&http.Request{RemoteAddr: "1.2.3.4:5555"},
			"5.6.7.8", "", "", "",
			http.Header{"Forwarded": {"for=1.2.3.4; proto=http; by=5.6.7.8"}},
			"",
		},

		{ // set Forwarded with localIP and HTTPS
			&http.Request{RemoteAddr: "1.2.3.4:5555", TLS: &tls.ConnectionState{}},
			"5.6.7.8", "", "", "",
			http.Header{"Forwarded": {"for=1.2.3.4; proto=https; by=5.6.7.8"}},
			"",
		},

		{ // extend Forwarded with localIP
			&http.Request{RemoteAddr: "1.2.3.4:5555", Header: http.Header{"Forwarded": {"for=9.9.9.9; proto=http; by=8.8.8.8"}}},
			"5.6.7.8", "", "", "",
			http.Header{"Forwarded": {"for=9.9.9.9; proto=http; by=8.8.8.8; by=5.6.7.8"}},
			"",
		},

		{ // set tls header
			&http.Request{RemoteAddr: "1.2.3.4:5555", TLS: &tls.ConnectionState{}},
			"5.6.7.8", "", "Secure", "",
			http.Header{"Forwarded": {"for=1.2.3.4; proto=https; by=5.6.7.8"}, "Secure": {""}},
			"",
		},

		{ // set tls header with value
			&http.Request{RemoteAddr: "1.2.3.4:5555", TLS: &tls.ConnectionState{}},
			"5.6.7.8", "", "Secure", "true",
			http.Header{"Forwarded": {"for=1.2.3.4; proto=https; by=5.6.7.8"}, "Secure": {"true"}},
			"",
		},
	}

	for i, tt := range tests {
		if tt.r.Header == nil {
			tt.r.Header = http.Header{}
		}

		err := addHeaders(tt.r, tt.localIP, tt.remoteIPHeader, tt.tlsHeader, tt.tlsHeaderValue)
		if err != nil {
			if got, want := err.Error(), tt.err; got != want {
				t.Errorf("%d: got %q want %q", i, got, want)
			}
			continue
		}
		if tt.err != "" {
			t.Errorf("%d: got nil want %q", i, tt.err)
			continue
		}
		if got, want := tt.r.Header, tt.hdrs; !reflect.DeepEqual(got, want) {
			t.Errorf("%d: got %v want %v", i, got, want)
		}
	}
}


main.go
package main

import (
	"flag"
	"fmt"
	"log"
	"net"
	"net/http"
	"os"
	"runtime"
	"runtime/debug"
	"strings"

	"github.com/eBay/fabio/consul"
	"github.com/eBay/fabio/metrics"
	"github.com/eBay/fabio/route"
	"github.com/eBay/fabio/ui"
)

var version = "1.0.4-issue-8-ssl-client-auth"

func main() {
	var cfg string
	var v bool
	flag.StringVar(&cfg, "cfg", "", "path to config file")
	flag.BoolVar(&v, "v", false, "show version")
	flag.Parse()

	if v {
		fmt.Println(version)
		return
	}

	log.Printf("[INFO] Version %s starting", version)

	if cfg != "" {
		if err := loadConfig(cfg); err != nil {
			log.Fatal("[FATAL] ", err)
		}
	}

	if err := metrics.Init(metricsTarget, metricsPrefix, metricsInterval, metricsGraphiteAddr); err != nil {
		log.Fatal("[FATAL] ", err)
	}

	if os.Getenv("GOMAXPROCS") == "" {
		log.Print("[INFO] Setting GOMAXPROCS=", gomaxprocs)
		runtime.GOMAXPROCS(gomaxprocs)
	} else {
		log.Print("[INFO] Using GOMAXPROCS=", os.Getenv("GOMAXPROCS"), " from env")
	}

	if os.Getenv("GOGC") == "" {
		log.Print("[INFO] Setting GOGC=", gogc)
		debug.SetGCPercent(gogc)
	} else {
		log.Print("[INFO] Using GOGC=", os.Getenv("GOGC"), " from env")
	}

	if proxyRoutes == "" {
		useDynamicRoutes()
	} else {
		useStaticRoutes()
	}

	if err := route.SetPickerStrategy(proxyStrategy); err != nil {
		log.Fatal("[FATAL] ", err)
	}

	consul.Addr = consulAddr
	consul.URL = consulURL

	dc, err := consul.Datacenter()
	if err != nil {
		log.Fatal("[FATAL] ", err)
	}

	log.Printf("[INFO] Using routing strategy %q", proxyStrategy)
	log.Printf("[INFO] Connecting to consul on %q in datacenter %q", consulAddr, dc)
	log.Printf("[INFO] Consul can be reached via %q", consulURL)

	log.Printf("[INFO] UI listening on %q", uiAddr)
	go func() {
		if err := ui.Start(uiAddr, consulKVPath); err != nil {
			log.Fatal("[FATAL] ui: ", err)
		}
	}()

	tr := &http.Transport{
		ResponseHeaderTimeout: proxyTimeout,
		MaxIdleConnsPerHost:   proxyMaxConn,
		Dial: (&net.Dialer{
			Timeout:   proxyDialTimeout,
			KeepAlive: proxyTimeout,
		}).Dial,
	}

	proxy := route.NewProxy(tr, proxyHeaderClientIP, proxyHeaderTLS, proxyHeaderTLSValue)
	listen(proxyAddr, proxyShutdownWait, proxy)
}

func useDynamicRoutes() {
	log.Printf("[INFO] Using dynamic routes from consul on %s", consulAddr)
	log.Printf("[INFO] Using tag prefix %q", consulTagPrefix)
	log.Printf("[INFO] Watching KV path %q", consulKVPath)
	go func() {
		w, err := consul.NewWatcher(consulTagPrefix, consulKVPath)
		if err != nil {
			log.Fatal("[FATAL] ", err)
		}
		w.Watch()
	}()
}

func useStaticRoutes() {
	var err error
	var t route.Table

	if strings.HasPrefix(proxyRoutes, "@") {
		proxyRoutes = proxyRoutes[1:]
		log.Print("[INFO] Using static routes from ", proxyRoutes)
		t, err = route.ParseFile(proxyRoutes)
	} else {
		log.Print("[INFO] Using static routes from config file")
		t, err = route.ParseString(proxyRoutes)
	}

	if err != nil {
		log.Fatal("[FATAL] ", err)
	}

	route.SetTable(t)
}


listen.go
package main

import (
	"crypto/tls"
	"crypto/x509"
	"io/ioutil"
	"log"
	"net"
	"net/http"
	"os"
	"os/signal"
	"regexp"
	"syscall"
	"time"

	"github.com/eBay/fabio/route"
)

var quit = make(chan bool)
var commas = regexp.MustCompile(`\s*,\s*`)
var semicolons = regexp.MustCompile(`\s*;\s*`)

func init() {
	go func() {
		// we use buffered to mitigate losing the signal
		sigchan := make(chan os.Signal, 1)
		signal.Notify(sigchan, os.Interrupt, os.Kill, syscall.SIGTERM)
		<-sigchan
		close(quit)
	}()
}

// listen starts one or more listeners for the handler. The list
// of addresses are
func listen(addrs string, wait time.Duration, h http.Handler) {
	for _, addr := range commas.Split(addrs, -1) {
		if addr == "" {
			continue
		}

		p := semicolons.Split(addr, 5)
		switch len(p) {
		case 1:
			go listenAndServe(p[0], h)
		case 2:
			go listenAndServeTLS(p[0], p[1], p[1], "", h)
		case 3:
			go listenAndServeTLS(p[0], p[1], p[2], "", h)
		case 4:
			go listenAndServeTLS(p[0], p[1], p[2], p[3], h)
		default:
			log.Fatal("[FATAL] Invalid address format ", addr)
		}
	}

	// wait for shutdown signal
	<-quit

	// disable routing for all requests
	route.Shutdown()

	// trigger graceful shutdown
	log.Printf("[INFO] Graceful shutdown over %s", wait)
	time.Sleep(wait)
	log.Print("[INFO] Down")
}

func listenAndServe(addr string, h http.Handler) {
	log.Printf("[INFO] HTTP proxy listening on %s", addr)
	if err := http.ListenAndServe(addr, h); err != nil {
		log.Fatal("[FATAL] ", err)
	}
}

// listenAndServeTLS starts an HTTPS server with the given certificate.
func listenAndServeTLS(addr, certFile, keyFile, clientAuthFile string, h http.Handler) {
	log.Printf("[INFO] HTTPS proxy listening on %s with certificate %s", addr, certFile)
	cert, err := tls.LoadX509KeyPair(certFile, keyFile)
	if err != nil {
		log.Fatal("[FATAL] ", err)
	}

	config := &tls.Config{
		NextProtos:   []string{"http/1.1"},
		Certificates: []tls.Certificate{cert},
	}

	if clientAuthFile != "" {
		pemBlock, err := ioutil.ReadFile(clientAuthFile)
		if err != nil {
			log.Fatal("[FATAL] ", err)
		}
		pool := x509.NewCertPool()
		if !pool.AppendCertsFromPEM(pemBlock) {
			log.Fatal("[FATAL] failed to add client auth certs")
		}
		config.ClientCAs = pool
		config.ClientAuth = tls.RequireAndVerifyClientCert
		log.Printf("[INFO] Client certificate authentication enabled on %s with certificates from %s", addr, clientAuthFile)
	}

	srv := &http.Server{Addr: addr, TLSConfig: config, Handler: h}

	ln, err := net.Listen("tcp", addr)
	if err != nil {
		log.Fatal("[FATAL] ", err)
	}

	tlsListener := tls.NewListener(tcpKeepAliveListener{ln.(*net.TCPListener)}, config)
	if err := srv.Serve(tlsListener); err != nil {
		log.Fatal("[FATAL] ", err)
	}
}

// copied from http://golang.org/src/net/http/server.go?s=54604:54695#L1967
// tcpKeepAliveListener sets TCP keep-alive timeouts on accepted
// connections. It's used by ListenAndServe and ListenAndServeTLS so
// dead TCP connections (e.g. closing laptop mid-download) eventually
// go away.
type tcpKeepAliveListener struct {
	*net.TCPListener
}

func (ln tcpKeepAliveListener) Accept() (c net.Conn, err error) {
	tc, err := ln.AcceptTCP()
	if err != nil {
		return
	}
	tc.SetKeepAlive(true)
	tc.SetKeepAlivePeriod(3 * time.Minute)
	return tc, nil
}


route/proxy.go
package route

import (
	"errors"
	"log"
	"net"
	"net/http"
	"net/http/httputil"
	"time"

	gometrics "github.com/eBay/fabio/_third_party/github.com/rcrowley/go-metrics"
)

// Proxy is a dynamic reverse proxy.
type Proxy struct {
	tr             http.RoundTripper
	localIP        string
	clientIPHeader string
	tlsHeader      string
	tlsHeaderValue string
	requests       gometrics.Timer
}

func NewProxy(tr http.RoundTripper, localIP, clientIPHeader, tlsHeader, tlsHeaderValue string) *Proxy {
	return &Proxy{
		tr:             tr,
		localIP:        localIP,
		clientIPHeader: clientIPHeader,
		tlsHeader:      tlsHeader,
		tlsHeaderValue: tlsHeaderValue,
		requests:       gometrics.GetOrRegisterTimer("requests", gometrics.DefaultRegistry),
	}
}

func (p *Proxy) ServeHTTP(w http.ResponseWriter, req *http.Request) {
	if ShuttingDown() {
		http.Error(w, "shutting down", http.StatusServiceUnavailable)
		return
	}

	target := GetTable().lookup(req, req.Header.Get("trace"))
	if target == nil {
		log.Print("[WARN] No route for ", req.URL)
		w.WriteHeader(404)
		return
	}

	if err := addHeaders(req, p.localIP, p.clientIPHeader, p.tlsHeader, p.tlsHeaderValue); err != nil {
		http.Error(w, "cannot parse "+req.RemoteAddr, http.StatusInternalServerError)
		return
	}

	start := time.Now()
	rp := httputil.NewSingleHostReverseProxy(target.URL)
	rp.Transport = p.tr
	rp.ServeHTTP(w, req)
	target.timer.UpdateSince(start)
	p.requests.UpdateSince(start)
}

func addHeaders(r *http.Request, localIP, clientIPHeader, tlsHeader, tlsHeaderValue string) error {
	remoteIP, _, err := net.SplitHostPort(r.RemoteAddr)
	if err != nil {
		return errors.New("cannot parse " + r.RemoteAddr)
	}

	if clientIPHeader != "" {
		r.Header.Set(clientIPHeader, remoteIP)
	}

	xff := r.Header.Get("X-Forwarded-For")
	if xff != "" && localIP != "" {
		r.Header.Set("X-Forwarded-For", xff+", "+localIP)
	}

	fwd := r.Header.Get("Forwarded")
	if fwd == "" {
		fwd = "for=" + remoteIP
		if r.TLS != nil {
			fwd += "; proto=https"
		} else {
			fwd += "; proto=http"
		}
	}
	if localIP != "" {
		fwd += "; by=" + localIP
	}
	r.Header.Set("Forwarded", fwd)

	if tlsHeader != "" && r.TLS != nil {
		r.Header.Set(tlsHeader, tlsHeaderValue)
	}

	return nil
}


config.go
package main

import (
	"net"
	"runtime"
	"time"

	"github.com/eBay/fabio/_third_party/github.com/magiconair/properties"
)

var (
	proxyAddr           = ":9999"
	proxyLocalIP        = localIP()
	proxyMaxConn        = 10000
	proxyRoutes         = ""
	proxyStrategy       = "rnd"
	proxyShutdownWait   = time.Duration(0)
	proxyDialTimeout    = 30 * time.Second
	proxyTimeout        = time.Duration(0)
	proxyHeaderClientIP = ""
	proxyHeaderTLS      = ""
	proxyHeaderTLSValue = ""
	consulAddr          = "localhost:8500"
	consulKVPath        = "/fabio/config"
	consulTagPrefix     = "urlprefix-"
	consulURL           = "http://" + consulAddr + "/"
	metricsTarget       = ""
	metricsInterval     = 30 * time.Second
	metricsPrefix       = "default"
	metricsGraphiteAddr = ""
	gogc                = 800
	gomaxprocs          = runtime.NumCPU()
	uiAddr              = ":9998"
)

func loadConfig(filename string) error {
	p, err := properties.LoadFile(filename, properties.UTF8)
	if err != nil {
		return err
	}

	proxyAddr = p.GetString("proxy.addr", proxyAddr)
	proxyLocalIP = p.GetString("proxy.localip", proxyLocalIP)
	proxyMaxConn = p.GetInt("proxy.maxconn", proxyMaxConn)
	proxyRoutes = p.GetString("proxy.routes", proxyRoutes)
	proxyStrategy = p.GetString("proxy.strategy", proxyStrategy)
	proxyShutdownWait = p.GetParsedDuration("proxy.shutdownWait", proxyShutdownWait)
	proxyDialTimeout = p.GetParsedDuration("proxy.dialtimeout", proxyDialTimeout)
	proxyTimeout = p.GetParsedDuration("proxy.timeout", proxyTimeout)
	proxyHeaderClientIP = p.GetString("proxy.header.clientip", proxyHeaderClientIP)
	proxyHeaderTLS = p.GetString("proxy.header.tls", proxyHeaderTLS)
	proxyHeaderTLSValue = p.GetString("proxy.header.tls.value", proxyHeaderTLSValue)
	consulAddr = p.GetString("consul.addr", consulAddr)
	consulKVPath = p.GetString("consul.kvpath", consulKVPath)
	consulTagPrefix = p.GetString("consul.tagprefix", consulTagPrefix)
	consulURL = p.GetString("consul.url", "http://"+consulAddr+"/")
	metricsTarget = p.GetString("metrics.target", metricsTarget)
	metricsInterval = p.GetParsedDuration("metrics.interval", metricsInterval)
	metricsPrefix = p.GetString("metrics.prefix", metricsPrefix)
	metricsGraphiteAddr = p.GetString("metrics.graphite.addr", metricsGraphiteAddr)
	gogc = p.GetInt("runtime.gogc", gogc)
	gomaxprocs = p.GetInt("runtime.gomaxprocs", gomaxprocs)
	uiAddr = p.GetString("ui.addr", uiAddr)

	return nil
}

// localIP tries to determine a non-loopback address for the local machine
func localIP() string {
	addrs, err := net.InterfaceAddrs()
	if err != nil {
		return ""
	}
	for _, addr := range addrs {
		if ipnet, ok := addr.(*net.IPNet); ok && ipnet.IP.IsGlobalUnicast() {
			if ipnet.IP.To4() != nil || ipnet.IP.To16() != nil {
				return ipnet.IP.String()
			}
		}
	}
	return ""
}
